### PR TITLE
use own arm64 gha runners

### PR DESCRIPTION
**Pull request by cicdteam**
## Problem

Move from aws based arm64 runners to bare-metal based

## Summary of changes
Changes in GitHub action workflows where `runs-on: arm64` used. More parallelism added, build time for `neon with extra platform builds` workflow reduced from 45m to 25m

### CODE REVIEW

File: .github/workflows/build-build-tools-image.yml [hunk 1]
```diff
@@ -39,7 +39,7 @@ jobs:
       matrix:
         arch: [ x64, arm64 ]
 
-    runs-on: ${{ fromJson(format('["self-hosted", "dev", "{0}"]', matrix.arch)) }}
+    runs-on: ${{ fromJson(format('["self-hosted", "large", "{0}"]', matrix.arch)) }}
```
Comment (bayandin) [inline .github/workflows/build-build-tools-image.yml:42]: We don't run this workflow if there are no changes in `Dockerfile.build-tools`; let's add some no-op changes there (or update some packages we have) to trigger this workflow.

File: .github/workflows/neon_extra_builds.yml [hunk 2]
```diff
@@ -232,20 +232,20 @@ jobs:
 
       - name: Run cargo build
         run: |
-          mold -run cargo build --locked $CARGO_FLAGS $CARGO_FEATURES --bins --tests
+          mold -run cargo build --locked $CARGO_FLAGS $CARGO_FEATURES --bins --tests -j$(nproc)
```
Comment (bayandin) [inline .github/workflows/neon_extra_builds.yml:235]: Should we add the same for our regular `.github/workflows/build_and_test.yml` workflow?

File: .github/workflows/neon_extra_builds.yml [hunk 3]
```diff
@@ -338,7 +347,7 @@ jobs:
       contains(github.event.pull_request.labels.*.name, 'run-extra-build-stats') ||
       contains(github.event.pull_request.labels.*.name, 'run-extra-build-*') ||
       github.ref_name == 'main'
-    runs-on: [ self-hosted, gen3, large ]
+    runs-on: [ self-hosted, large, x64 ]
```
Comment (bayandin) [inline .github/workflows/neon_extra_builds.yml:350]: Is this a new label for the current x86-64 runners we have? I'd either update it in all the places or keep it as is

File: .github/workflows/neon_extra_builds.yml [hunk 4]
```diff
@@ -269,6 +269,11 @@ jobs:
         password: ${{ secrets.NEON_DOCKERHUB_PASSWORD }}
       options: --init
 
+    strategy:
+      fail-fast: false
+      matrix:
+        build_type: [ debug, release ]
+
```
Comment (bayandin) [inline .github/workflows/neon_extra_builds.yml:276]: Are these changes necessary for the PR?

File: .github/actionlint.yml [hunk 5]
```diff
@@ -4,10 +4,10 @@ self-hosted-runner:
     - dev
     - gen3
     - large
-    # Remove `macos-14` from the list after https://github.com/rhysd/actionlint/pull/392 is merged.
-    - macos-14
     - small
     - us-east-2
+    - ubuntu-22.04-arm64
+    - ubuntu-latest-arm64
```
[the remark this answers is not in the document]
Comment (cicdteam) [reply .github/actionlint.yml:10]: yes, that was the main idea - to use almost the same names so that you can easily move from GitHub-hosted to self-hosted runners. I mean just replace `runs-on: ubuntu-latest` with `runs-on: [self-hosted, ubuntu-latest]` to use our more powerful runners. Do you think this is a bad idea?

File: .github/workflows/neon_extra_builds.yml (hunk 2 above)
Comment (cicdteam) [reply .github/workflows/neon_extra_builds.yml:235]: By default, the cargo should use the maximum number of available cores. But I always prefer to specify the number of jobs. Currently I've only worked with the workflow associated with Arm64 builds, but I might check out other workflows (preferably in a separate PR).

File: .github/workflows/neon_extra_builds.yml (hunk 3 above)
Comment (cicdteam) [reply .github/workflows/neon_extra_builds.yml:350]: yes, `x64` and `arm64` are labels inherited automatically from uname info (and `Linux` as well)

File: .github/workflows/neon_extra_builds.yml (hunk 4 above)
Comment (cicdteam) [reply .github/workflows/neon_extra_builds.yml:276]: this reduced the workflow time from 45 minutes to 25 minutes as these jobs now run in parallel.

File: .github/actionlint.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (bayandin) [reply .github/actionlint.yml:10]: I think it's better to give them more distinguishable label because GitHub's runners are ephemeral, and self-hosted runners are long-running, so we need to be more careful with self-hosted runners to not break them for the next jobs

[the remark this answers is not in the document]
Comment (cicdteam) [reply .github/actionlint.yml]: label changed to `large-arm64`

File: .github/workflows/build-build-tools-image.yml (hunk 1 above)
Comment (cicdteam) [reply .github/workflows/build-build-tools-image.yml]: tested (added one \r at the end of dockerfile). Built with success.

File: .github/workflows/build-build-tools-image.yml [hunk 6]
```diff
@@ -39,7 +39,7 @@ jobs:
       matrix:
         arch: [ x64, arm64 ]
 
-    runs-on: ${{ fromJson(format('["self-hosted", "dev", "{0}"]', matrix.arch)) }}
+    runs-on: ${{ fromJson(format('["self-hosted", "gen3", "{0}"]', matrix.arch)) }}
```
Comment (bayandin) [inline .github/workflows/build-build-tools-image.yml:42]: Won't it choose a random runner for x86-64 (small/large)?

Should we use something like this:
```suggestion
    runs-on: ${{ fromJson(format('["self-hosted", "gen3", "{0}"]', matrix.arch == 'arm64' && 'large-arm' || 'large')) }}
```

File: Dockerfile.build-tools [hunk 7]
```diff
@@ -170,3 +170,4 @@ RUN whoami \
     && rustup --version --verbose \
     && rustc --version --verbose \
     && clang --version
+
```
Comment (bayandin) [inline Dockerfile.build-tools:173]: And we can revert this, probably

**Review by bayandin — APPROVED** [review]
Great work, thanks Andrey!
